pkg/source/ttnv2/util.go

```diff
@@ -0,0 +1,12 @@
+package ttnv2
+
+func unmarshalWithTool(
+	tool interface {
+		UnmarshalText([]byte) error
+		Bytes() []byte
+	},
+	source string,
+) ([]byte, error) {
+	err := tool.UnmarshalText([]byte(source))
+	return tool.Bytes(), err
+}
```

Comment:
I don't get this. What is `tool` about a target variable?

Reply:
The interface could be named `unmarshaller` instead, the tool name is left after [this change](https://github.com/TheThingsNetwork/lorawan-stack-migrate/compare/980edb779030198bcd812177dfbd16fb59b08829..308e624726a0d9027810b9b2863ab01182b283b8). The interface is used to unmarshall keys in a cleaner manner

Reply:
But the "tool" has state. Reusing the tool changes the bytes there were returned earlier.

What is the benefit over just using `UnmarshalText()` directly?

Reply:
Since the changes in newer `lorawan-stack` versions (`v3.15.2` -> `v3.21`) each block like
```go

Reply:
Since the changes in newer `lorawan-stack` versions (`v3.15.2` -> `v3.21`) each block
```go
dev.RootKeys.AppKey = &ttnpb.KeyEnvelope{
  Key: &types.AES128Key{},
}
if err := dev.RootKeys.AppKey.Key.UnmarshalText([]byte(rootKeys.AppKey)); err != nil {
  return nil, errInvalidKey.WithAttributes(rootKeys.AppKey).WithCause(err)
}
```
would need to be changed to
```go

dev.RootKeys.AppKey = &ttnpb.KeyEnvelope{}
appKey := &types.AES128{}
if err := appKey.UnmarshalText(rootKeys.AppKey.Bytes()); err != nil {
  return nil, errInvalidKey.WithAttributes(rootKeys.AppKey).WithCause(err)
}
dev.RootKeys.AppKey.Key = appKey.Bytes()
```
and using `unmarshalWithTool()` it looks like
```go
dev.RootKeys.AppKey = &ttnpb.KeyEnvelope{}
dev.RootKeys.AppKey.Key, err = unmarshalWithTool(&types.AES128{}, rootKeys.AppKey)
if err != nil {
  return nil, errInvalidKey.WithAttributes(rootKeys.AppKey).WithCause(err)
}
```
which seemed cleaner to me. Each time a new tool can be declared the same as with `UnmarshalText()` approach, just doesn't need a separate line and no need to assign to a value again.

Those code fragments tended to be repeated multiple times one below another thus seemed beneficial to remove the clutter.

Reply:
If the approach is correct i think the `unmarshalWithTool()` could be upgraded with use of generics
```go
func unmarshalToBytes[T interface {
  UnmarshalText([]byte) error
  Bytes() []byte
}](
  source T,
) ([]byte, error) {
  var tool T
  err := tool.UnmarshalText(source.Bytes())
  return tool.Bytes(), err
}
```
to get rid of the "tool"
```go
dev.RootKeys.AppKey = &ttnpb.KeyEnvelope{}
dev.RootKeys.AppKey.Key, err = unmarshalToBytes(rootKeys.AppKey)
if err != nil {
  return nil, errInvalidKey.WithAttributes(rootKeys.AppKey).WithCause(err)
}
```

Reply:
This wins one line of code, right? I still don't see the real benefit, can you explain?

Reply:
Just a dozen lines of code in both packages, no more benefit.
Should those be changed to the second (middle) method?

Reply:
Removed completely from `pkg/source/ttnv2` and renamed in `pkg/source/chirpstack`